api/src/Controller/IncludeResource.php

```diff
@@ -0,0 +1,81 @@
+<?php declare(strict_types=1);
+/*.
+    require_module 'standard';
+.*/
+
+namespace App\Controller;
+
+use Slim\Container;
+use Slim\Http\Request;
+use Slim\Http\Response;
+
+class IncludeResource
+{
+
+    /**
+     * @var string
+     */
+    private $class;
+
+    /**
+     * @var string
+     */
+    private $field;
+
+    /**
+     * @var string
+     */
+    private $parameter;
+
+
+    public function __construct(string $class, string $parameter, string $field)
+    {
+        $this->class = $class;
+        $this->field = $field;
+        $this->parameter = $parameter;
+
+    }
+
+
+    public function process(Request $request, Response $response, Container $container, array $params, array &$data, array $history): void
+    {
+        if (in_array($this->field, array_keys($data), true) &&
```

Comment:
Yeah it is a pretty dense little recursive function.  When I get a some cycles I will try to help you unwind it.

Reply:
```
    public function process(Request $request, Response $response, Container $container, array $params, array &$data, array $history): void
    {
        /* Verify this include field is in the resulting data, and make sure that the data is not null */
        if (in_array($this->field, array_keys($data), true) &&
            $data[$this->field] !== null) {
            /* Check if this resource class is in the history and if not add a blank history entry for this class */
            if (!array_key_exists($this->class, $history)) {
                $history[$this->class] = [];
            }
            /* check if in the class history this field has already been looked up. If so we do not continue */
            if (in_array($data[$this->field], $history[$this->class])) {
                return;
            }
            /* Build the parameters to call into the included resource */
            $newparams = $params;
            $newparams[$this->parameter] = $data[$this->field];
            $target = new $this->class($container);
            /* Call the buildResource for the included resource and if we crash just get out */
            try {
                $newdata = $target->buildResource($request, $response, $newparams)[1];
            } catch (Exception $e) {
                return;
            }
            /* Add this field to the history */
            $history[$this->class][] = $data[$this->field];
            /* Process all the includes for the newly included resource (Recursion) */
            IncludeResource::processIncludes($target->includes, $request, $response, $container, $params, $newdata, $history);
            /* Add all the response to the data for the resource */
            $data[$this->field] = $target->arrayResponse($request, $response, $newdata);
        }
```
Inlined comments because that was easiest for me to do.  I can include these comments if you feel like it makes it clearer.

Reply:
The comments are helpful, and it confirms that I was on the right track. I think it would still be nice if we broke it into smaller functions though because it will give clarity through function names about what is being done and then we don't have to add comments to describe what is happening.

I'll let you decide which way to go with it. I'm including some example pseudocode for how it could be broken out into a smaller function just in case.

As a quick example, here's some pseudocode for what I mean:

```
public function process(Request $request, Response $response, Container $container, array $params, array &$data, array $history)
{
  if (hasIncludeData($data)) {
    initalizeHistory($history);
    if (hasExistingHistory($data, $history)) {
       return;
    }

    try {
       collectResourceData($request, $response, $container, $params, $data);
       $history[$this->class][] = $data[$this->field];
       processNextInclude($target->includes, $request, $response, $container, $params, $data, $newData, $history);
    } catch (Exception $e) {
       return;
    }
  }
}

private function hasIncludeData(array $data): boolean
{
  // Check if array has key and actually has data value
}

private function initalizeHistory(array $history): void
{
  // Check if the history exists and add it if not
}

private function hasExistingHistory(array $data, array $history): boolean
{
  // Check if the history already existed
}

private function collectResourceData(Request $request, Response $response, Container $container, array $params, array $data)
{
   // Include lines 50-54, do not include try/catch
}

private function processNextInclude ($includes, Request $request, Response $response, Container $container, array $params, array &$data, array $newdata, array $history): void
{
  IncludeResource::processIncludes($includes, $request, $response, $container, $params, $newdata, $history);
  $data[$this->field] = $target->arrayResponse($request, $response, $newdata);
}
```